MaintenanceJobDesignDoc.md

```diff
@@ -0,0 +1,140 @@
+Maintenance Job Design Document
+===============================
+
+The (shared) object cache is growing to be large. This includes:
+
+* Number of loose objects.
+* Number of pack-files.
+* Size of pack-files.
+
+There are some things we can do to improve this situation, with varying levels
+of difficulty and benefit. Here are the four items I believe will have a strong
+impact on our data size and performance:
+
+I. Delete loose objects that exist in pack-files.
+II. Place loose objects into pack-files.
+III. Repack some prefetch packs into larger packs.
+IV. Delete packs with no objects referenced by the multi-pack-index.
+
+These are numbered by order of implementation, but they work together in an
+interesting way. If we place loose objects into pack-files (I), then the next
+time we delete loose objects (II) we will delete those objects that we just put
+into a pack-file. If we repack some prefetch packs into larger packs (III), the
+multi-pack-index will use the "newer" pack for all objects in those packs, and
+we can delete them (IV).
+
+In order to perform this maintenance without disrupting users, we can actually

```

Comment:
I split the loose objects and prefetch pack maintenance into separate jobs just because the timing involved seems like it should be different. Running A.1, A.2, B.1, B.2 (using numbers from below) in order will not disrupt users. However, if we did A.2 _then_ A.1, we may disrupt users, because they may try to access a loose object as we delete it.